File: src/bugbot/services/diff.py
"""Unified-diff parser. Just enough to map each added line back to its
line number in the post-change file, so inline comments land correctly.

We do NOT use a third-party diff lib — keeping the dependency surface
small matters in a tool that may be installed into CI images."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from fnmatch import fnmatch

_FILE_HEADER_RE = re.compile(r"^diff --git a/(?P<old>.+?) b/(?P<new>.+?)$")
_HUNK_HEADER_RE = re.compile(
    r"^@@ -(?P<old_start>\d+)(?:,(?P<old_count>\d+))? "
    r"\+(?P<new_start>\d+)(?:,(?P<new_count>\d+))? @@"
)


@dataclass
class DiffLine:
    new_lineno: int | None  # None for deletions
    old_lineno: int | None  # None for additions
    kind: str  # "+", "-", " "
    content: str


@dataclass
class DiffHunk:
    new_start: int
    new_count: int
    old_start: int
    old_count: int
    lines: list[DiffLine] = field(default_factory=list)


@dataclass
class FileDiff:
    old_path: str
    new_path: str
    is_new: bool = False
    is_deleted: bool = False
    is_binary: bool = False
    hunks: list[DiffHunk] = field(default_factory=list)

    @property
    def path(self) -> str:
        return self.new_path if not self.is_deleted else self.old_path

    def added_line_numbers(self) -> list[int]:
        return [
            ln.new_lineno
            for h in self.hunks
            for ln in h.lines
            if ln.kind == "+" and ln.new_lineno is not None
        ]

    def added_lines(self) -> list[tuple[int, str]]:
        return [
            (ln.new_lineno, ln.content)
            for h in self.hunks
            for ln in h.lines
            if ln.kind == "+" and ln.new_lineno is not None
        ]


def _strip_prefix(path: str) -> str:
    if path.startswith(("a/", "b/")):
        return path[2:]
    return path


def parse_unified_diff(diff: str) -> list[FileDiff]:
    files: list[FileDiff] = []
    current: FileDiff | None = None
    current_hunk: DiffHunk | None = None
    new_lineno = 0
    old_lineno = 0

    for raw in diff.splitlines():
        # --- file header ---
        m = _FILE_HEADER_RE.match(raw)
        if m:
            if current is not None:
                files.append(current)
            current = FileDiff(
                old_path=_strip_prefix(m.group("old")),
                new_path=_strip_prefix(m.group("new")),
            )
            current_hunk = None
            continue

        if current is None:
            continue

        if raw.startswith("new file mode"):
            current.is_new = True
            continue
        if raw.startswith("deleted file mode"):
            current.is_deleted = True
            continue
        if raw.startswith("Binary files") or raw.startswith("GIT binary patch"):
            current.is_binary = True
            continue
        if raw.startswith(("--- ", "+++ ", "index ", "similarity ", "rename ")):
            continue

        # --- hunk header ---
        m = _HUNK_HEADER_RE.match(raw)
        if m:
            current_hunk = DiffHunk(
                new_start=int(m.group("new_start")),
                new_count=int(m.group("new_count") or 1),
                old_start=int(m.group("old_start")),
                old_count=int(m.group("old_count") or 1),
            )
            current.hunks.append(current_hunk)
            new_lineno = current_hunk.new_start
            old_lineno = current_hunk.old_start
            continue

        if current_hunk is None:
            continue

        if not raw:
            # blank line inside hunk is a context line with empty content
            current_hunk.lines.append(DiffLine(
                new_lineno=new_lineno, old_lineno=old_lineno, kind=" ", content=""
            ))
            new_lineno += 1
            old_lineno += 1
            continue

        kind = raw[0]
        content = raw[1:]
        if kind == "+":
            current_hunk.lines.append(DiffLine(
                new_lineno=new_lineno, old_lineno=None, kind="+", content=content
            ))
            new_lineno += 1
        elif kind == "-":
            current_hunk.lines.append(DiffLine(
                new_lineno=None, old_lineno=old_lineno, kind="-", content=content
            ))
            old_lineno += 1
        elif kind == " ":
            current_hunk.lines.append(DiffLine(
                new_lineno=new_lineno, old_lineno=old_lineno, kind=" ", content=content
            ))
            new_lineno += 1
            old_lineno += 1
        elif kind == "\\":
            # "\ No newline at end of file" — ignore.
            continue

    if current is not None:
        files.append(current)
    return files


def filter_files(files: list[FileDiff], ignore_globs: list[str]) -> list[FileDiff]:
    if not ignore_globs:
        return files
    kept: list[FileDiff] = []
    for f in files:
        path = f.path
        if any(fnmatch(path, g) for g in ignore_globs):
            continue
        kept.append(f)
    return kept

File: src/services/diff.rs
//! Unified-diff parser — just enough to map each added line back to its line
//! number in the post-change file, so inline comments land correctly. Ported
//! from the Python `services/diff.py`; no third-party diff lib by design.

use std::sync::LazyLock;

use globset::{Glob, GlobSetBuilder};
use regex::Regex;

static FILE_HEADER_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^diff --git a/(.+?) b/(.+?)$").expect("file header re"));

static HUNK_HEADER_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@").expect("hunk header re")
});

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiffLine {
    pub new_lineno: Option<u32>,
    pub old_lineno: Option<u32>,
    pub kind: char, // '+', '-', ' '
    pub content: String,
}

#[derive(Debug, Clone, Default)]
pub struct DiffHunk {
    pub new_start: u32,
    pub new_count: u32,
    pub old_start: u32,
    pub old_count: u32,
    pub lines: Vec<DiffLine>,
}

#[derive(Debug, Clone, Default)]
pub struct FileDiff {
    pub old_path: String,
    pub new_path: String,
    pub is_new: bool,
    pub is_deleted: bool,
    pub is_binary: bool,
    pub hunks: Vec<DiffHunk>,
}

impl FileDiff {
    fn new(old_path: String, new_path: String) -> Self {
        FileDiff {
            old_path,
            new_path,
            ..Default::default()
        }
    }

    /// The path to anchor comments / scan against (post-change, unless this is
    /// a deletion).
    pub fn path(&self) -> &str {
        if self.is_deleted {
            &self.old_path
        } else {
            &self.new_path
        }
    }

    pub fn added_line_numbers(&self) -> Vec<u32> {
        self.hunks
            .iter()
            .flat_map(|h| h.lines.iter())
            .filter(|l| l.kind == '+')
            .filter_map(|l| l.new_lineno)
            .collect()
    }

    /// (new_lineno, content) for every added (`+`) line.
    pub fn added_lines(&self) -> Vec<(u32, &str)> {
        self.hunks
            .iter()
            .flat_map(|h| h.lines.iter())
            .filter(|l| l.kind == '+')
            .filter_map(|l| l.new_lineno.map(|n| (n, l.content.as_str())))
            .collect()
    }
}

fn strip_prefix(path: &str) -> String {
    path.strip_prefix("a/")
        .or_else(|| path.strip_prefix("b/"))
        .unwrap_or(path)
        .to_string()
}

pub fn parse_unified_diff(diff: &str) -> Vec<FileDiff> {
    let mut files: Vec<FileDiff> = Vec::new();
    let mut current: Option<FileDiff> = None;
    // Index of the hunk currently being filled within `current`.
    let mut in_hunk = false;
    let mut new_lineno: u32 = 0;
    let mut old_lineno: u32 = 0;

    for raw in diff.lines() {
        // --- file header ---
        if let Some(caps) = FILE_HEADER_RE.captures(raw) {
            if let Some(done) = current.take() {
                files.push(done);
            }
            current = Some(FileDiff::new(
                strip_prefix(&caps[1]),
                strip_prefix(&caps[2]),
            ));
            in_hunk = false;
            continue;
        }

        let Some(file) = current.as_mut() else {
            continue;
        };

        if raw.starts_with("new file mode") {
            file.is_new = true;
            continue;
        }
        if raw.starts_with("deleted file mode") {
            file.is_deleted = true;
            continue;
        }
        if raw.starts_with("Binary files") || raw.starts_with("GIT binary patch") {
            file.is_binary = true;
            continue;
        }
        if raw.starts_with("--- ")
            || raw.starts_with("+++ ")
            || raw.starts_with("index ")
            || raw.starts_with("similarity ")
            || raw.starts_with("rename ")
        {
            continue;
        }

        // --- hunk header ---
        if let Some(caps) = HUNK_HEADER_RE.captures(raw) {
            let old_start = caps[1].parse().unwrap_or(0);
            let old_count = caps
                .get(2)
                .and_then(|m| m.as_str().parse().ok())
                .unwrap_or(1);
            let new_start = caps[3].parse().unwrap_or(0);
            let new_count = caps
                .get(4)
                .and_then(|m| m.as_str().parse().ok())
                .unwrap_or(1);
            file.hunks.push(DiffHunk {
                new_start,
                new_count,
                old_start,
                old_count,
                lines: Vec::new(),
            });
            in_hunk = true;
            new_lineno = new_start;
            old_lineno = old_start;
            continue;
        }

        if !in_hunk {
            continue;
        }
        let hunk = file
            .hunks
            .last_mut()
            .expect("in_hunk implies a hunk exists");

        if raw.is_empty() {
            // A bare blank line inside a hunk = a context line with empty content.
            hunk.lines.push(DiffLine {
                new_lineno: Some(new_lineno),
                old_lineno: Some(old_lineno),
                kind: ' ',
                content: String::new(),
            });
            new_lineno += 1;
            old_lineno += 1;
            continue;
        }

        let kind = raw.as_bytes()[0] as char;
        let content = raw[1..].to_string();
        match kind {
            '+' => {
                hunk.lines.push(DiffLine {
                    new_lineno: Some(new_lineno),
                    old_lineno: None,
                    kind: '+',
                    content,
                });
                new_lineno += 1;
            }
            '-' => {
                hunk.lines.push(DiffLine {
                    new_lineno: None,
                    old_lineno: Some(old_lineno),
                    kind: '-',
                    content,
                });
                old_lineno += 1;
            }
            ' ' => {
                hunk.lines.push(DiffLine {
                    new_lineno: Some(new_lineno),
                    old_lineno: Some(old_lineno),
                    kind: ' ',
                    content,
                });
                new_lineno += 1;
                old_lineno += 1;
            }
            // "\ No newline at end of file" and anything else: ignore.
            _ => {}
        }
    }

    if let Some(done) = current.take() {
        files.push(done);
    }
    files
}

/// Drop files matching any ignore glob. `*` matches across `/` (fnmatch
/// semantics), so `*.lock` matches `a/b/c.lock` and `vendor/**` matches
/// anything under `vendor/`. Invalid globs are skipped (logged).
pub fn filter_files(files: Vec<FileDiff>, ignore_globs: &[String]) -> Vec<FileDiff> {
    if ignore_globs.is_empty() {
        return files;
    }
    let mut builder = GlobSetBuilder::new();
    for g in ignore_globs {
        match Glob::new(g) {
            Ok(glob) => {
                builder.add(glob);
            }
            Err(e) => tracing::warn!("ignoring invalid glob {g:?}: {e}"),
        }
    }
    let set = match builder.build() {
        Ok(s) => s,
        Err(e) => {
            tracing::warn!("could not build ignore globset: {e}; keeping all files");
            return files;
        }
    };
    files
        .into_iter()
        .filter(|f| !set.is_match(f.path()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    const SAMPLE: &str = "\
diff --git a/src/app.py b/src/app.py
index 1111111..2222222 100644
--- a/src/app.py
+++ b/src/app.py
@@ -1,3 +1,4 @@
 import os
-x = 1
+x = 2
+y = 3
 print(x)
diff --git a/notes.txt b/notes.txt
new file mode 100644
--- /dev/null
+++ b/notes.txt
@@ -0,0 +1,2 @@
+hello
+world
";

    #[test]
    fn parses_two_files() {
        let files = parse_unified_diff(SAMPLE);
        assert_eq!(files.len(), 2);
        assert_eq!(files[0].path(), "src/app.py");
        assert!(files[1].is_new);
        assert_eq!(files[1].path(), "notes.txt");
    }

    #[test]
    fn added_lines_have_correct_new_numbers() {
        let files = parse_unified_diff(SAMPLE);
        // app.py: context "import os" (1), -x=1, +x=2 (2), +y=3 (3), context print (4)
        let added: Vec<(u32, &str)> = files[0].added_lines();
        assert_eq!(added, vec![(2, "x = 2"), (3, "y = 3")]);
        assert_eq!(files[0].added_line_numbers(), vec![2, 3]);
        // notes.txt new file: lines 1,2
        assert_eq!(files[1].added_line_numbers(), vec![1, 2]);
    }

    #[test]
    fn binary_marker_detected() {
        let d = "diff --git a/img.png b/img.png\nBinary files a/img.png and b/img.png differ\n";
        let files = parse_unified_diff(d);
        assert!(files[0].is_binary);
    }

    #[test]
    fn filter_files_respects_globs() {
        let files = parse_unified_diff(SAMPLE);
        let globs = vec!["*.txt".to_string()];
        let kept = filter_files(files, &globs);
        assert_eq!(kept.len(), 1);
        assert_eq!(kept[0].path(), "src/app.py");
    }

    #[test]
    fn filter_files_star_matches_across_slash() {
        let d = "diff --git a/vendor/lib/x.js b/vendor/lib/x.js\n@@ -0,0 +1,1 @@\n+code\n";
        let files = parse_unified_diff(d);
        let kept = filter_files(files, &["vendor/**".to_string()]);
        assert!(kept.is_empty());
    }
}
